L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,470 @@
+  // not including the root.
+  absl::string_view peer_cert_full_chain() const;
+
+  // The common name string from the Subject extension in the peer's leaf
+  // certificate.
+  absl::string_view common_name() const;
+
+  // The subject name of the root certificate used to verify the peer chain
+  // If verification fails or the peer certificate is self-signed, this will be an
+  // empty string. If verification is successful, it is a comma-separated list,
+  // where the entries are of the form "FIELD_ABBREVIATION=string"
+  // ex: "CN=testca,O=Internet Widgits Pty Ltd,ST=Some-State,C=AU"
+  // ex: "CN=GTS Root R1,O=Google Trust Services LLC,C=US"
+  absl::string_view verified_root_cert_subject() const;
+
+  // The (possibly empty) list of URI names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  absl::Span<absl::string_view> uri_names() const;
+
+  // The (possibly empty) list of DNS names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  absl::Span<absl::string_view> dns_names() const;
+
+  // The (possibly empty) list of email names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  absl::Span<absl::string_view> email_names() const;
+
+  // The (possibly empty) list of IP names from the Subject Alternative Name
+  // extension in the peer's leaf certificate.
+  absl::Span<absl::string_view> ip_names() const;
+}
+
+// The base class of all verifier implementations. Note that custom
+// verifier implementations can compose their functionality with existing
+// implementations of this interface, such as HostNameVerifier, by delegating
+// to an instance of that class.
+class CertificateVerifierInterface {
+ public:
+  virtual ~CertificateVerifierInterface() = default;
+
+  // Performs a custom verification check.
+  // Returns true if verification occurs synchronously, in which case the
+  // verification result is populated in the sync_status output parameter.
+  // Returns false if verification occurs asynchronously, in which case the
+  // callback must be called when verification is complete and the verification
+  // result must be populated in the sync_status output parameter.
+  //
+  // request: the verification information associated with this request
+  // callback: This will only take effect if the verifier is asynchronous.
+  //           The function that gRPC will invoke when the verifier has already
+  //           completed its asynchronous check. Callers can use this function
+  //           to perform any additional checks. The input parameter of the
+  //           std::function indicates the status of the verifier check.
+  // return: 
+  // If error status, failed synchronously.
+  // If OK status and true, succeeded synchronously.
+  // If OK status and false, verification is still in progress asynchronously.
+  virtual absl::StatusOr<bool> Verify(VerifiedPeerCertificateChainInfo* request,
```

Comment:
micro-nit: I don't think the variable name should still be "request".

Reply:
Changed to `verified_chain_info`